### ISSUE TITLE
Start of music is cut off on start screen of Bumpy's Arcade Fantasy

### ISSUE THREAD

**Issue by madbrain76** [munt/munt#119]
I should preface this by saying that I no longer own a physical MT-32/CM-32L/LAPC-1 .I had all of them at one point when I was younger, but not anymore. I definitely did play the game Bumpy on my LAPC-1 a little over 30 years ago. However, I just cannot recall if this problem existed on the real thing or not after so much time.

The game can be downloaded from https://d2.myabandonware.com/t/a3a0e547-1d78-44c1-8f74-e8004badd897/Bumpys-Arcade-Fantasy_DOS_EN.zip and runs well under both Dosbox and Dosbox-x. The MT-32 problem is the same.

Anyway, I captured the video of the startup screen of the game with Dosbox-x. 
https://www.youtube.com/watch?v=R3rFI-TpSuY
It is very obvious - at least to me - that at 0:03, a note or two are missing.
The same problem occurs in Dosbox too (non-X) with Munt library 2.7.0 .

So, I took it upon myself to record the MIDI output of the game using Dosbox.

[bumpy.zip](https://github.com/munt/munt/files/15480806/bumpy.zip)

I did several tests with this MIDI file..
1) used the Tools / Play MIDI file function. The start was still cut off.
2) used the MIDI-Ox Midibar program to play the file on the Munt device. The start was cut off, too.
3) opened it in MIDI editor, installed from http://www.midieditor.org/ . And there, it showed all the notes from the start, and played them, without a cut !
4) I also used MIDI-Ox Midibar to play the file on my physical Roland Fantom-XR over USB. To my dismay, the starting notes were cut off, too.

It could just be a bug in the game, not giving enough time between the opening of the device before starting playback. All the abstraction layers in a modern OS, it's driver architecture, and a VM, could be causing a tiny bit of extra latency. Only someone with the original device on bare metal on DOS could really say for sure. And I don't have any of the required hardware left. Perhaps someone in the enthusiast community still does, though.

Nevertheless, the missing notes are actually in the MIDI file I recorded, and didn't get played, so perhaps there is a way to work around it ? Unless of course it is determined the real device under DOS has the problem too, and bug compatibility is desired.

**Comment by madbrain76** [comment]
The game "Baby Jo" has the same problem with the startup music. The game must be started with "BABY.EXE 3" to get MT-32 sound. Download link is https://d2.myabandonware.com/t/2d74218c-de84-4bdf-a922-01762eadec5c/Baby-Jo-in-Going-Home_DOS_EN.zip . I definitely remember cracking its copy protection when it came out, though I didn't sign my work and it could be someone else's crack in there. 

Anyway, I attached the recording.
[babyjo.zip](https://github.com/munt/munt/files/15481334/babyjo.zip)

The results were slightly different :

1. used the Tools / Play MIDI file function. The start was still cut off.
2. used the MIDI-Ox Midibar program to play the file on the Munt device. The start was cut off, too.
3. I opened it in MIDI editor, and it showed all the notes from the start, and played them all, without a cut !
4. I also used MIDI-Ox Midibar to play the file on my physical Roland Fantom-XR over USB. To my surprise, the starting notes played fine (other than being all the wrong patches). They were not cut off, unlike with Bumpy.

The publisher for these 2 games is the same, Loriciel. It's conceivable the programmers shared code.

**Comment by johnnovak** [comment]
@madbrain76 Can you post a screenshot of the missing notes in question? Are the notes overlapping, perhaps? E.g., two note on messages followed by two note off messages later (so Note On, Note On, Note Off, Note Off on the same channel for the the same note, instead of Note On, Note Off, Note On, Note Off).

**Comment by madbrain76** [comment]
@johnnovak ,
Not quite sure which program to use to dump the MIDI events, much less try to do a cut in the middle of the playback. But visually, with MIDI editor on Windows, it looks like this :

Notice the mouse pointer near the top, and the red vertical line. That's the point at which the music starts in my Youtube video. Anything to the left of that is skipped.

![image](https://github.com/munt/munt/assets/57333073/b7efc93b-a1a4-4676-b7b4-d90867835ebb)

**Comment by johnnovak** [comment]
Hmm, I think it's a different issue then. Could it be MT-32 ROM revision dependent? Does the game send SysEx messages at the start? It could also be something related to that perhaps. E.g. the game sending the init SysEx messages, then the actual MIDI notes follow that "too quickly". Perhaps try with really low cycles settings in DOSBox, e.g. `cycles = 1000` or `2000`, maybe that changes something.

**Comment by johnnovak** [comment]
Okay, I've tested it too and played around with the DOSBox settings, but nothing improves the situation. Bit of a mystery!

**Comment by madbrain76** [comment]
Yeah, same results with cycles = 1000 . Really not sure what's going on . Would be helpful to test it on a real device as opposed to emulator, if somebody has one handy.

FYI, I opened Bumpy.mid in Seikaju 8.1 . It shows the same problem - first few notes don't sound (almost silent, not 100%) - but same notes missing when playing the game. Unfortunately, the patches are all wrong, pretty much all piano. There are probably Sysex in there.

MIDI editor plays this file correctly, without missing notes, and correct-sounding patches.

Neither program handles all recorded MIDI files from DOSbox fully. Ran into many problems with MIDI editor. Have not tried most with Seikaju which I just installed. Also, my list of recorded game MIDs is growing by the day.

**Comment by johnnovak** [comment]
@madbrain76 Yeah, I'm pretty sure it's a SysEx-related init weirdness or latency issue.

**Comment by madbrain76** [comment]
Yeah, very possible. I wonder if there is a way to work around the timing problem. There is only one timing-related setting, which is the MIDI delay mode. Perhaps there needs to be something more granular ? Maybe just do processing differently at the start of a MIDI stream (say, after MIDI has been quiet for a while). Just throwing ideas there. I'm not sure what could really fix this.
But I think hooking up a physical MT-32 and using it in DOSbox with these games would be a good start to find a solution.

**Comment by sergm** [comment]
I believe, it is a timing problem. When I tried setting "MIDI Delay Mode" -> "Process all MIDI events immediately", the beginning of `bumpy.mid` sounded more-or-less complete. IMO, the game didn't delay the initial burst of kinda "reset" events, so the timing of the subsequent events got screwed. This evil looks like sending NoteOff for every possible note on every single part, all up to address `0x11F2`. All these events come with delay 0. Note, there are no SysExes whatsoever.

Ultimately, the problem here is caused by buffering all those numerous events in the reset pack. In the software MIDI path, they get to the synth instantly, and by default, mt32emu takes into account the time they'd need to get to a real hardware unit via a real MIDI cable. This results in delaying the subsequent events by this time, so the beginning appears "compressed".

Of course, one'd prefer sending a short SysEx message that faithfully resets the synth instead of such nonsense. But adding a deliberate delay after the last message from the pack `91 01 00` would also work, I suppose.

Speaking of a potential way to fix this in DOSBox, we could enforce delaying short MIDI events similarly to how SysEx events are delayed already, when `delaysysex` parameter is added to `midiconfig`.

**Comment by johnnovak** [comment]
Thanks @sergm, this was very informative. I've played around with delay modes in DOSBox Staging, and `setMIDIDelayMode(MT32Emu::MIDIDelayMode_IMMEDIATE)` fixes the problem: the notes are not cut at the start in this mode, and also the SysEx message is not ignored so some of the instruments sound different.

`DELAY_SHORT_MESSAGES_ONLY` only fixes the cut notes, but the SysEx somehow doesn't get correctly processed, so the instruments aren't set up correctly.

I'm not quite sure if DOSBox does something wrong, or the MT-32 driver is a bit broken on real hardware as well. I can't test it as I don't have a hardware MT-32. However, I'm quite certain the sounds I'm getting in "immediate" mode are the correct ones, as they are similar to the Amiga version of the soundtrack:

https://www.youtube.com/watch?v=hju8oJdzvo8

Note the xylophone like sounds—you only get those in immediate delay mode.

I'm inclined to make a change to DOSBox Staging to always set the immediate delay mode. In your experience, does this have any drawbacks? Are you aware of any cases where the accurate MIDI transfer speed emulation would make a noticeable difference and that difference is desirable? I get it that the transfer speed emulation could introduce micro-delays between notes and events which are gone in immediate mode, but I'm hard-pressed to think this is important. Losing the start of the tune and not setting up custom patches properly via SysEx is a much bigger problem IMO.

**Comment by madbrain76** [comment]
@johnnovak ,
For me, setting immediate mode in Munt does not fully fix the problem in Bumpy. It may be cut off very slightly less, but still cutoff.
For Babyjo, I got 3 different results :
1) funny note (very sharp) at the start . typed F10 after the home screen to exit
2) went to PC speaker instead f MIDI when typing BABY 3
3) on the 3rd try, everything played fine
So, I would say even immediate mode is not the full fix.

**Comment by johnnovak** [comment]
> For me, setting immediate mode in Munt does not fully fix the problem in Bumpy. It may be cut off very slightly less, but still cutoff.

Yeah, I can hear that. I've made a recording of Staging's output using immediate mode, and yeah, the first one or two notes are slightly cut. You can hear the correct version at 1:23 where the tune loops back to the start.

https://drive.google.com/file/d/1gKdodauGtvFbQvurIDrM8GhIf-WKAS3k/view?usp=sharing

Not sure how to improve this further, but "immediate mode" seems to be a significant improvement if it has no major drawbacks (even it not perfect, but now we're very close).

Again, the game might not play the tune perfectly even on real hardware for all we know 🤷🏻

I will experiment with Babyjo tomorrow if I'll have some time.

**Comment by madbrain76** [comment]
I have been playing a couple other games in immediate mode. One of them is Prehistorik. On one of the levels, some of patches were wrong. I switched back to the default mode, and didn't have this problem any longer. But I don't know for sure if this is related to the mode or not. I wasn't recording when it happened. So far, when I record, I am doing only MIDI recordings, not PCM. They are quite hard to edit, especially in a way that you can loop them. I guess that's the nature of MIDI. Some have markers for loop start/end, but they aren't defined in the SMF standard ...

**Comment by johnnovak** [comment]
Yeah, it's hard to know whether the bug was introduced by immediate mode or not, unless you play the game several times in both modes. Also, unless something sounds obviously wrong, it's hard to tell what is wrong.

I will try a variety of games in immediate mode too to see if I can notice any differences.

In the worst case, immediate mode could be another config option, but we're trying hard to use defaults that need no tweaking is possible. It would be nice if we could just hardcode immediate mode if it doesn't cause problems.

Sorry @sergm for hijacking your issue tracker, we'll stop and continue in the DOSBox Staging tracker with @madbrain76 from now 😄

**Comment by madbrain76** [comment]
I am mainly playing the games I already know. Prehistorik was one of them. There is also already a video on Youtube of the whole soundtrack. I could tell from the first measure that something was off.

I had not gotten the issue with the games running in PC speaker mode before instead of MIDI. I just tried BabyJo with default settings, and it happened too - exactly like I describe in this comment : https://github.com/munt/munt/issues/119#issuecomment-2143810152 .

So, that part is not related

For Prehistorik I would need to try again.

**Comment by sergm** [comment]
Well, there is obviously one more problem with both games we initially considered. The thing is, the initial burst of NoteOn messages may also lead to an overflow of the internal MIDI queue. The latter appeared to have capacity of 1024 by default, and this is about the number of offending MIDI events...

In fact, if I play either of the MIDI files via an external MIDI player, the interim processing delay is sufficient for mt32emu-qt to clear some space in the queue. But when I use the built-in player, the sounding will very likely be off. Once again: there are no SysEx messages within, _at all_. Still, right after the silly NoteOn sequence, a series of Program Change events come, which might happen to be discarded in case the queue is already full.

Looking at the roots of these problems, we should consider what is the best way of handling backpressure. Now mt32emu-qt just discards events if they come in way too fast. Unfortunately, it's a tricky business to propagate backpressure upstream, to the application that emits MIDI events. Some MIDI drivers may simply delay the caller process while an event is on its way to the target device. Others do buffering and discard extra events, just like mt32emu-qt does. But even if we put e.g. the DOSBox process on hold, it's by far not obvious how the original DOS application will react on that. It might delay the subsequent events, or it might also discard missed ones, who knows...

**Comment by johnnovak** [comment]
Thanks for the clarification @sergm. It's still unclear to me how this game can work flawlessly on hardware, but probably it can't. In any case, just focusing on the buffer overflow problem, why not oversize the buffer? E.g. raise it to 4096 or 8192 from 1024? I don't think any MIDI synth does backpressure; the MIDI protocop has no such feature. We just need a bigger buffer to drop messages later. The amount of "burst" MIDI messages is limited on every hardware, so we just need a bigger buffer to make the issue less likely to happen. Personally, I'd use 8182 and move on.

**Comment by sergm** [comment]
As I said, the behaviour on a real hardware depends on the driver. When a DOS game runs inside DOSBox, there are actually **two drivers**: one implemented in the game and another runs on the host OS. Each of them may buffer quickly coming events, and none of them is guaranteed to have sufficient capacity. I admit, those games likely work smoothly when both drivers wait for each MIDI event to be delivered via the MIDI interface, which is basically ~1 millisecond. Hence, I proposed that analog of `delaysysex` config option but for short events, so that DOSBox might prevent the need to handle backpressure at the level of the OS driver and downstream. Because, whatever buffer size we configure, we'll surely find a use case where it is not enough 😄

**Comment by sergm** [comment]
Hmm, I am actually reluctant to raise the buffer size. 1024 is AFAICT a tiny bit more than the size of the inner buffer new MT-32 has allocated. The biggest problem with having a huge buffer is the potential for lags. We might buffer up the entire tune, and new events will be processed after a few minutes. There are certainly games which play FX via MIDI, or do some other realtime things.

If we get back to our poor games, we already have a 1-second lag at start. Important to keep in mind that MIDI events are buffered along with their timestamps, to ensure desired event timing of playback. Events that are too old are played back ASAP, to shorten the apparent lag anyway, even if they don't get discarded.

**Comment by johnnovak** [comment]
@sergm Okay, we've brought up a fair number of theories and "what if" scenarios, I will just test this on hardware today, capture the MIDI stream, and then compare it with the MIDI stream output by DOSBox Staging. I'd really like to get to the bottom of this 😄 In particular, I'm curious if the "instantaneous" and "bursty" message packets get "smeared out" a bit in the time domain on real hardware at the pure raw MIDI output level.

Also found this interesting summary of buffer sizes on various MPU implementations:
https://www.vogons.org/viewtopic.php?p=597742#p597742

Here's the image and the notes:

![image](https://github.com/munt/munt/assets/698770/74529ddf-8b7d-43e3-82b9-4251f1b26266)

> - Transmit Buffer = Host to MPU
> - Receive Buffer = External MIDI Input to MPU
> - Populated fields represent information that is documented, mostly in datasheets and programming references, where the fields highlighted in green have been personally determined and/or verified.
> - Mega-Em employs software-based circular-buffering. I'm not sure what the impact on the transmit buffer is, if any, but it bumps the receive buffer from 16-bytes (hardware) to 255-bytes.
> - Windows' MMSYSTEM.DLL, while not represented in or by this data, also seems to employ circular-buffering.
> - Where a dedicated, non-MCU-based UART is used, 1-byte buffering is adequate (the Atari ST being another such example).

Just purely restricting this exercise to the DOS side of things (let's not even get into the modern host OS, API, and MIDI device behaviour)

- Whatever the DOS game is doing — probably just outputting bytes to the MPU in UART mode at precisely timed intervals (there's no timestamping at that level, the MIDI stream needs to be sent in real-time as precisely as desired/possible)
- Some potential buffering at the MPU Transmit Buffer level as per the above diagram. This is just 1 byte for Creative cards (no buffer), 8 bytes for ESS cards, and 256 bytes for the classic Roland MPU-401. Well, I can only test on an ESS and an AWE32 myself.
- The internal MIDI stream buffering of the MT-32 itself which is a 1024 buffer ring buffer in your case (MT32EMU_DEFAULT_MIDI_EVENT_QUEUE_SIZE). The MIDI protocol doesn't do timestamping, so you must timestamp events in the Munt code as you're receiving them in real-time.

So, I see two places where the "spreading out" (or "throttling") of "bursty" MIDI events can happen which might cause the MT-32 not to overflow its internal input even buffer with this game:

- **DOS -> MPU:** In the DOS game's code, which for some reason might not be emulated accurately (that's a DOSBox problem if true). Maybe in DOSBox we're not emulating some latencies between the successive MIDI send messages as it happens on hardware from the DOS code to the MPU.
- **MPU -> MT-32:** Some timing delays could be introduced by the MPU implementation too, somehow. Probably the behaviour is slightly different on every MPU on different sound card brands...

Anyway, I will run some tests and report back. I will only be able to determine the combined delay (if any) for the whole `**DOS -> MPU ** MT-32** path—that's easy enough, and I'm not willing to go deeper at this point 😅  Then we'll have some more data points based on actual measurements of actual hardware.

Enabling "immediate mode" is the easy solution for Staging with our integrated libmt32, and I'm quite certain the resulting microscopiC timing differences are not significant and could only be noticed under laboratory conditions 😎 But I'm slightly worried how things will behave with Nuked SC55 once we integrate it; that project accurately simulates the hardware, so enabling "immediate mode" might not be an option, and then we nee

**Comment by johnnovak** [comment]
> @johnnovak , For me, setting immediate mode in Munt does not fully fix the problem in Bumpy. It may be cut off very slightly less, but still cutoff. For Babyjo, I got 3 different results :
> 
>     1. funny note (very sharp) at the start . typed F10 after the home screen to exit
> 
>     2. went to PC speaker instead f MIDI when typing BABY 3
> 
>     3. on the 3rd try, everything played fine
>        So, I would say even immediate mode is not the full fix.

I've been able to reproduce this exactly, it's very strange. I wonder if the game is just badly coded... I will test this game on real hardware as well and capture the MIDI output, maybe that will shed some light on the issue.

**Comment by sergm** [comment]
I need to say, DOS hardware wasn't perfectly uniform (as the table above clearly shows), and testing on all sorts of hardware is what we former DOS users did in the time :smile:

No surprise, if the game works well in some configurations but fails in another. I am pretty sure there are some. I owned (in the past life) a MIDI-Tech USB-MIDI cable which has a buffer of a few tens bytes, and when that buffer overflows, the damn thing simply discarded upcoming MIDI bytes. So I even wrote a driver for it that calls `sleep` to hold the emitter. OTOH, a cheap USB-MIDI cable with 4-byte buffer worked flawlessly just because it made the standard driver wait until the USB packet gets sent.

But yeah, let's see what your tests tell us :100:

**Comment by johnnovak** [comment]
> I need to say, DOS hardware wasn't perfectly uniform (as the table above clearly shows), and testing on all sorts of hardware is what we former DOS users did in the time 😄
> 
> No surprise, if the game works well in some configurations but fails in another. I am pretty sure there are some. I owned (in the past life) a MIDI-Tech USB-MIDI cable which has a buffer of a few tens bytes, and when that buffer overflows, the damn thing simply discarded upcoming MIDI bytes. So I even wrote a driver for it that calls `sleep` to hold the emitter. OTOH, a cheap USB-MIDI cable with 4-byte buffer worked flawlessly just because it made the standard driver wait until the USB packet gets sent.
> 
> But yeah, let's see what your tests tell us 💯

Yeah, it's definitely hard to test these things "authoritatively" 😅 

In my case, the chain will be:

Pentium MMX downclocked to 100 Mhz -> ESS 1868F / AWE 32 -> Kenton Thru-5 (MIDI splitter box) -> M-Audio MIDISPORT2x2 (USB) -> some laptop running Windows 7 -> record in REAPER

So yeah, quite a few stages between the source and the final destination 😅

**Comment by madbrain76** [comment]
I don't think I ever got any MIDI game to work on the joystick port of the various SBs I owned over the years. However, the LAPC-1 and later MQX-16 both worked flawlessly as far as I can remember until the Pentium 100 I owned in 1996. After that, PCI started replacing VLB and ISA.

**Comment by johnnovak** [comment]
> I don't think I ever got any MIDI game to work on the joystick port of the various SBs I owned over the years. However, the LAPC-1 and later MQX-16 both worked flawlessly as far as I can remember until the Pentium 100 I owned in 1996. After that, PCI started replacing VLB and ISA.

Well, most people playing games had a "consumer" board with built-in MPU after they became common, so some ESS/ Creative or similar board and they used the joystick port.

But yeah, in MT-32 times, more people had the Roland MPU cards/boxes, that's probably correct.

**Comment by sergm** [comment]
I'm not sure whether the zip files the OP posted are created by DOSBox or by some other means. I did analyse them though, and came up with what I told above. Technically, it is not possible to send a thousand MIDI events in a few MIDI ticks which are equal to 1 millisecond over a serial MIDI interface, so the capture does suggest for some incorrectness or unfairness ;)

**Comment by madbrain76** [comment]
I definitely created them with Dosbox. .now, which fork of it, I'm not 100% certain. Happy to take another capture after I sober up and get out of the hot tub..

**Comment by johnnovak** [comment]
> I definitely created them with Dosbox. .now, which fork of it, I'm not 100% certain. Happy to take another capture after I sober up and get out of the hot tub..

I loaded it into a sequencer, but I didn't see thousands of messages. Probably the sequencer does it's own cleanup or something. I guess I'll need to inspect the raw MIDI file then somehow. What did you use for that @sergm ?

**Comment by sergm** [comment]
>  I didn't see thousands of messages

Just a wild guess: it might be all those get collapsed into one as they have 0 delay in between.

**Comment by johnnovak** [comment]
Ok, so I've added some trace logging to Staging, and indeed, the game spams the MPU with 1000+ Note Off messages for all the 128 MIDI notes on the first 9 channels. That's a very silly thing to do...

I have a theory that a real MPU-401 starts dropping events once its buffers are full, thus effectively "protecting" the MT-32. Then not all Note Offs get through on real hardware, but that doesn't really matter on a freshly inited MT-32, so the music just plays fine.

The more I think about, the more I'm inclined to just set immediate mode in Munt and move on. That seems to be the most compatible behaviour; dropping events arbitrarily when there is a surge of events seem risky without understanding the exact intelligent mode MPU-401 behaviour. Things can get really murky there... Maybe the MPU generated data request interrupts are more spread out on real hardware because of bus speeds and the MPU-401 isn't a CRAY supercomputer either 😄 

So, rather inconclusive, but we have a practical fix. And yeah, I don't trust the MPU-401 intelligent mode implementation too much. In fact, now I'm feeling motivated to fix it after 20 years so it finally plays those Legend Entertainment MT-32 pieces without weird tempo fluctuations...

**Comment by sergm** [comment]
Interesting though: what does it do when the buffer overflows? It could:

- discard the new event
- drop the oldest event
- put a lock on the bus and wait

**Comment by johnnovak** [comment]
Actually, the third option seems plausible. The MPU generates interrupts per channel to request the next midi event (up to 4 bytes, the first one is the optional timing info which is the whole point of using intelligent mode on an XT or 286). Theorethically, the MPU could backpressure the DOS program this way.

I will generate further traces at the MPU protocol level as looking at the MIDI events emitted by DOSBox is only half the story.

I would bet good money on it the buffer overload scenario of the MPU is not emulated at all but they just did something "simple".

**Comment by madbrain76** [comment]
Here are new recordings done with DOSBox Staging 0.81.1 . Munt 2.7.0 was in the default mode (delay short messages only).

[Capture.zip](https://github.com/user-attachments/files/15533831/Capture.zip)

**Comment by sergm** [comment]
> Actually, the third option seems plausible. The MPU generates interrupts per channel to request the next midi event (up to 4 bytes, the first one is the optional timing info which is the whole point of using intelligent mode on an XT or 286). Theorethically, the MPU could backpressure the DOS program this way.

I also think that if MPU discarded some events it wouldn't be that intelligent. But it seems that "delaying the sender" is actually the simplest solution, and I bet the cheap "dumb" devices do exactly that. By the idea, we don't want to lose any data but might sacrifice the timing accuracy.

**Comment by madbrain76** [comment]
BTW, both of the games I mentioned - Bumpy and Baby Jo - have MIDI files in their distribution, in case this can help with the emulation. The title score for Baby Jo is in the file MUSIC8.MID . It plays with a high note at the start when using Munt's "Play MIDI file" feature.

**Comment by sergm** [comment]
Ah, thanks! I'll take a look at them.

**Comment by sergm** [comment]
Well, the original file does seem to have significant delays between the events in the initial sequence. Very sus...

**Comment by sergm** [comment]
> The title score for Baby Jo is in the file MUSIC8.MID . It plays with a high note at the start when using Munt's "Play MIDI file" feature.

This is according to the original. The very first NoteOn in the file is sent before Program Change, so a freshly turned on MT-32 will play it using a different timbre.

**Comment by sergm** [comment]
And the original score for Bumpy is even more interesting. The NoteOn messages there do not map 1:1 to Program Change events, so normal MIDI players and the one built in to mt32emu-qt fail to play it correctly. Although, it is fairly easy to remap the MIDI channels 1->2 in MIDI-Ox or any editor.

**Comment by madbrain76** [comment]
> > @johnnovak , For me, setting immediate mode in Munt does not fully fix the problem in Bumpy. It may be cut off very slightly less, but still cutoff. For Babyjo, I got 3 different results :
> > ```
> > 1. funny note (very sharp) at the start . typed F10 after the home screen to exit
> > 
> > 2. went to PC speaker instead f MIDI when typing BABY 3
> > 
> > 3. on the 3rd try, everything played fine
> >    So, I would say even immediate mode is not the full fix.
> > ```
> 
> I've been able to reproduce this exactly, it's very strange. I wonder if the game is just badly coded... I will test this game on real hardware as well and capture the MIDI output, maybe that will shed some light on the issue.
> 
> I've tried immediate mode, different cycles settings, `vmem_delay`, `core = simple`, and nothing has improved the situation.

Could very well be Baby Jo was badly coded. But it might also be an MPU-401 emulation issue.

In any case, I dug out an old program I wrote in Pascal with inline assembly that resets both the MPU-401 and the MT-32.
If I run this program immediately before Baby Jo, I get the behavior mentioned above for #1, ie. the tune is cut off at the beginning. I attached the binary for this program. I remember using this program quite a bit way back on real hardware, especially the LAPC-1 which did not have any reset button
[reset.zip](https://github.com/user-attachments/files/15540418/reset.zip)
, unlike the MT-32. CM-32L also only had a volume button.

**Comment by johnnovak** [comment]
@madbrain76 Oh yeah, resetting the MT-32 is a must if you run one game after another is DOSBox. With a real hardware device you must power cycle it, or send it the reset commands, otherwise the game will often find the MT-32 in a random state. Many games don't reset it at start or when exiting.

With built-in Munt in Staging and using a single config per game setup this problem is sidestepped.

**Comment by madbrain76** [comment]
@johnnovak I'm using the external Munt for Windows so I can play MIDI files outside of DOSBox. 
The reset program I wrote seems to have an effect not just on the synth but on the MPU also. Without it, on the second attempt, Baby Jo fails to detect the MPU, and goes to PC speaker. I think we should probably take this issue to a separate bug.

**Comment by johnnovak** [comment]
> @johnnovak I'm using the external Munt for Windows so I can play MIDI files outside of DOSBox. The reset program I wrote seems to have an effect not just on the synth but on the MPU also. Without it, on the second attempt, Baby Jo fails to detect the MPU, and goes to PC speaker. I think we should probably take this issue to a separate bug.

I think it's a game bug, though. If the game doesn't reset the MPU on startup or exit, and leaves the MPU in a weird state on exit, this is not a DOSBox bug.

**Comment by madbrain76** [comment]
It's quite possible the game is buggy. I certainly can't recall restarting the game on real hardware or how it behaved on bare metal 30+ years later. It's also possible it's a timing-related MPU emulation issue. Is there any kind of log for emulated MPU commands, and if so where ? Same question for access to the PC speaker. The only thing I have seen so far in the status window are log entries about the SB DSP when it is used/probed.

**Comment by johnnovak** [comment]
> Is there any kind of log for emulated MPU commands, and if so where 

There isn't. You can add `LOG_TRACE` and `LOG_DEBUG` etc statements to `src/hardware/mpu401.cpp` yourself right now. I'm intending to do that at some point and ifdef them out of normal builds. In any case, you'll need to compile yourself a debug build to see those logs.

> Same question for access to the PC speaker. 

Same answer 😄

**Comment by sergm** [comment]
I can't disagree with the statement about the buggy games. I am most certain about their MIDI driver though: I seriously never ever seen such a reset sequence anywhere else. Sane people simply never do that. However, the original bundled MIDI files also do not seem perfect in my eyes, most probably because those were created using an unrelated sequencer.

**Comment by madbrain76** [comment]
@johnnovak , I think it would be worthwhile to have an option log file even in the release build, at least of the content of what already gets logged onto the console. If anything, it should help diagnose bug reports in the future, IMO.

I just haven't taken the time to get it to build yet. I did install Visual Studio 2022 CE.. Very weirdly, in step 2, there is no "Visual Studio Developer Command Prompt " for me to run after the installation. This is the second time I try to install it. I definitely have the C++ clang files. Maybe the third time will be the charm.

```
 Directory of C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\Llvm\x64\bin

06/03/2024  19:08       125,557,816 clang++.exe
```
Or maybe I can just use vcvars32.bat to run vcpkg - this is what I did.

BTW, I was ale to checkout the source for dosbox-staging from within Visual Studio 2022. No need to install git.exe .

It looks like my build has started successfully. It's downloading all the dependencies. I'll turn off power savings on my host and let it build.

**Comment by madbrain76** [comment]
As I stated in the comment below, I'm not seeing any change in behavior with Bumpy .

https://github.com/dosbox-staging/dosbox-staging/pull/3729#issuecomment-2148906809

**Comment by johnnovak** [comment]
> I think it would be worthwhile to have an option log file even in the release build, at least of the content of what already gets logged onto the console. If anything, it should help diagnose bug reports in the future, IMO.

It would have advantages, perhaps, but for troubleshooting issues, you generally need to insert manual trace logging anyway for the specific things you wanna test. It would be a _lot_ of work to add debug & trace logging everywhere and cover every scenario. Even after spending a lot of time on it, you might need to add more logging, so you'd need to rebuild anyway...

Also, we should probably stop (mis)using the Munt issue tracker for Staging related conversations 😄 Please join our Discord to chat about these things in realtime.

**Comment by sergm** [comment]
Closed due to inactivity.
